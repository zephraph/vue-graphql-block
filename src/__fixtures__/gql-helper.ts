import { readFile } from 'fs';
import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { join } from 'path';

import { splitDocument } from '../query-validators';

const rel = fileName => join(__dirname, 'gql', `${fileName}.gql`);

export const readQuery = (name: string): Promise<DocumentNode[]> =>
  new Promise((resolve, reject) => {
    readFile(rel(name), 'utf8', (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(splitDocument(gql(data) as DocumentNode));
    });
  });
