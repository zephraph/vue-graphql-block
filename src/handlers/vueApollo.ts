import { DocumentNode, OperationDefinitionNode, DefinitionNode } from 'graphql';
import Vue, { ComponentOptions } from 'vue';
import { GraphQLBlockAttributes } from '../index';

export const ANONYMOUS_QUERY = '__anon_query__';

const getDefinitionNode = (doc: DocumentNode): DefinitionNode =>
  doc.definitions[0];

const getQueryName = (
  docDef: OperationDefinitionNode,
  { alias }: GraphQLBlockAttributes = {}
) => {
  if (alias) return alias;
  if (docDef.name) return docDef.name.value;
  return ANONYMOUS_QUERY;
};

const isQuery = (definitionNode: DefinitionNode) =>
  definitionNode.kind === 'OperationDefinition' &&
  definitionNode.operation === 'query';

export const buildApolloOptions = (
  gqlDocuments: DocumentNode[],
  attributes: GraphQLBlockAttributes
) =>
  gqlDocuments
    .filter(doc => isQuery(getDefinitionNode(doc)))
    .map(doc => ({
      [getQueryName(
        getDefinitionNode(doc) as OperationDefinitionNode,
        attributes
      )]: {
        query: doc
      }
    }))
    .reduce((curr, acc) => ({ ...acc, ...curr }), {});

export type Handler = (
  component: {
    options: ComponentOptions<Vue>;
  },
  gqlDocuments: DocumentNode[],
  attributes: GraphQLBlockAttributes
) => void;

const vueApolloHandler: Handler = function handler(
  component,
  gqlDocuments,
  attributes = {}
) {
  component.options.apollo = {
    ...component.options.apollo,
    ...buildApolloOptions(gqlDocuments, attributes)
  };
};

export default vueApolloHandler;
