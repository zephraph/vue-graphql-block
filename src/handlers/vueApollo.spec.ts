import { strToDocNodes } from '../__fixtures__/gql-helper';
import { buildApolloOptions, ANONYMOUS_QUERY } from './vueApollo';

const basicNamedQuery = `query Hello { world }`;
const basicAnonQuery = `{ hello }`;

describe('Vue Apollo Handler', () => {
  describe('buildApolloOptions', () => {
    it('should match a simple named query snapshot', () => {
      const query = buildApolloOptions(strToDocNodes(basicNamedQuery), {});
      expect(query).toMatchSnapshot();
      expect(query).toMatchObject({ Hello: { query: {} } });
    });

    it('should match a simple anonymous query snapshot', () => {
      const query = buildApolloOptions(strToDocNodes(basicAnonQuery), {});
      expect(query).toMatchSnapshot();
      expect(query).toMatchObject({ [ANONYMOUS_QUERY]: { query: {} } });
    });
  });
});
