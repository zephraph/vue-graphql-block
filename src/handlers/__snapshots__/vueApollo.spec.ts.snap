// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue Apollo Handler buildApolloOptions should return a simple named query 1`] = `
Object {
  "Hello": Object {
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "Hello",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "world",
                },
                "selectionSet": undefined,
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 21,
        "start": 0,
      },
    },
  },
}
`;
